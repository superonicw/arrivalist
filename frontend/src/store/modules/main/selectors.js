import { get } from 'lodash'

export const selectRushingState = state => get(state, 'rushing')

export const selectRushingData = state => get(state, 'rushing.data')

export const selectTeams = state => get(state, 'rushing.teams')

export const selectRushingLoading = state => get(state, 'rushing.loading')

export const selectRushingError = state => get(state, 'rushing.error')
