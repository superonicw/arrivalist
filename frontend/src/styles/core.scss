div#root {
  height: 100%;
}

.linechart {
  width: 100%;
  height: 100%;
}

.mapchart {
  text {
    pointer-events: none;
  }
}
