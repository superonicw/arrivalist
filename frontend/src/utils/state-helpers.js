export const successAction = action => `${action}/success`

export const failAction = action => `${action}/fail`
