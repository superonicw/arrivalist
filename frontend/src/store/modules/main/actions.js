import { createAction } from 'redux-actions'
import { successAction, failAction } from 'utils/state-helpers'

export const LIST_RECORD = 'LIST_RECORD'

export const listRecord = createAction(LIST_RECORD)
export const listRecordSuccess = createAction(successAction(LIST_RECORD))
export const listRecordFail = createAction(failAction(LIST_RECORD))
