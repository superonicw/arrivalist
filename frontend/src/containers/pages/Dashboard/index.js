import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { pickBy, identity } from 'lodash'
import { makeStyles } from '@material-ui/core/styles'
import CssBaseline from '@material-ui/core/CssBaseline'
import AppBar from '@material-ui/core/AppBar'
import Button from '@material-ui/core/Button'
import Toolbar from '@material-ui/core/Toolbar'
import Typography from '@material-ui/core/Typography'
import Container from '@material-ui/core/Container'
import Grid from '@material-ui/core/Grid'
import Paper from '@material-ui/core/Paper'
import CircularProgress from '@material-ui/core/CircularProgress'
import { listTrip, selectTrips, selectLoading } from 'store/modules/main'
import { Filter, LineChart, MapChart } from 'components'

const useStyles = makeStyles(theme => ({
  root: {
    display: 'flex',
  },
  toolbar: {
    paddingRight: 24,
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  menuButton: {
    marginRight: 36,
  },
  title: {
    flexGrow: 1,
  },
  appBarSpacer: theme.mixins.toolbar,
  content: {
    flexGrow: 1,
    height: '100vh',
    overflow: 'auto',
  },
  container: {
    paddingTop: theme.spacing(4),
    paddingBottom: theme.spacing(4),
  },
  paper: {
    padding: theme.spacing(2),
    display: 'flex',
    overflow: 'auto',
    flexDirection: 'column',
  },
  button: {
    marginTop: theme.spacing(2),
  },
}))

const Dashboard = () => {
  const classes = useStyles()

  const [state, setState] = useState('')
  const [dates, setDates] = useState({ start: '', end: '' })

  const trips = useSelector(selectTrips)
  const loading = useSelector(selectLoading)

  const dispatch = useDispatch()

  function fetchData() {
    const { start, end } = dates

    if (!state && !start && !end) {
      return
    }

    const params = pickBy({ state, start, end }, identity)

    dispatch(listTrip({ params }))
  }

  return (
    <div className={classes.root}>
      <CssBaseline />
      <AppBar position="absolute" className={classes.appBar}>
        <Toolbar className={classes.toolbar}>
          <Typography
            component="h1"
            variant="h6"
            color="inherit"
            noWrap
            className={classes.title}
          >
            Dashboard
          </Typography>
        </Toolbar>
      </AppBar>

      <main className={classes.content}>
        <div className={classes.appBarSpacer} />
        <Container maxWidth={false} className={classes.container}>
          <Grid container spacing={3}>
            <Grid item xs={12} md={2}>
              <Paper className={classes.paper}>
                <Filter
                  state={state}
                  dates={dates}
                  onStateChange={setState}
                  onDatesChange={setDates}
                />
                <Button
                  variant="contained"
                  color="primary"
                  className={classes.button}
                  disabled={(!state && !dates.start && !dates.end) || loading}
                  onClick={fetchData}
                >
                  {loading ? (
                    <CircularProgress color="secondary" size={24} />
                  ) : (
                    'Search'
                  )}
                </Button>
              </Paper>
            </Grid>
            <Grid item xs={12} md={5}>
              <Paper className={classes.paper}>
                <LineChart width={700} height={548} trips={trips} />
              </Paper>
            </Grid>
            <Grid item xs={12} md={5}>
              <Paper className={classes.paper}>
                <MapChart trips={trips} />
              </Paper>
            </Grid>
          </Grid>
        </Container>
      </main>
    </div>
  )
}

export default Dashboard
