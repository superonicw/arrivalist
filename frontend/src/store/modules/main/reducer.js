import { handleActions } from 'redux-actions'
import { successAction, failAction } from 'utils/state-helpers'
import { LIST_TRIP, CLEAR_TRIP } from './actions'

export const INITIAL_STATE = {
  trips: [],
  loading: false,
  error: null,
}

export const reducer = handleActions(
  {
    [LIST_TRIP]: state => ({
      ...state,
      error: null,
      loading: true,
    }),

    [CLEAR_TRIP]: state => ({
      ...state,
      trips: [],
    }),

    [successAction(LIST_TRIP)]: (state, { payload }) => ({
      ...state,
      trips: payload,
      loading: false,
    }),

    [failAction(LIST_TRIP)]: (state, { payload }) => ({
      ...state,
      loading: false,
      error: payload,
    }),
  },
  INITIAL_STATE,
)
