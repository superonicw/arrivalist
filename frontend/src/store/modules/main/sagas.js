import axios from 'axios'
import { call, put, takeEvery } from 'redux-saga/effects'

import * as actions from './actions'

export const doListTrip = function* ({ payload }) {
  try {
    const res = yield call(axios.get, '/trips/', payload)
    yield put(actions.listTripSuccess(res.data.trips))
  } catch (error) {
    yield put(actions.listTripFail(error))
  }
}

export const saga = function* () {
  yield takeEvery(actions.LIST_TRIP, doListTrip)
}
