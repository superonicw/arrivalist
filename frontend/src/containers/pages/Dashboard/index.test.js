import React from 'react'
import { mount, shallow } from 'enzyme'
import { createStore } from 'redux'
import { set } from 'lodash'
import { Button, CircularProgress, Select, TextField } from '@material-ui/core'
import { listTrip, clearTrip } from 'store/modules/main'
import reducer from 'store/reducers'
import { withProvider } from 'test/utils'
import { TripMock } from 'test/mocks'
import Dashboard from './index'

const initialState = {
  main: {
    trips: TripMock,
    loading: false,
    error: null,
  },
}

describe('Dashboard page test', () => {
  let wrapper
  let store

  beforeEach(() => {
    store = createStore(reducer, initialState)
    store.dispatch = jest.fn()
    wrapper = withProvider(mount, <Dashboard />, store)
  })

  it('renders correctly', () => {
    expect(withProvider(shallow, <Dashboard />, store)).toMatchSnapshot()
  })

  it('should handle search', () => {
    const states = ['AK', 'AL']
    const start = '2020-08-01'

    wrapper
      .find(Select)
      .props()
      .onChange({ target: { value: states } })

    wrapper
      .find(TextField)
      .first()
      .props()
      .onChange({ target: { value: start } })

    wrapper.find(Button).last().simulate('click')

    const params = {
      states: states.join(','),
      start,
    }

    expect(store.dispatch).toHaveBeenLastCalledWith(listTrip({ params }))

    wrapper.find(Button).first().simulate('click')
    expect(store.dispatch).toHaveBeenLastCalledWith(clearTrip())
  })

  it('should show loader', () => {
    store = createStore(reducer, set({ ...initialState }, 'main.loading', true))
    wrapper = withProvider(mount, <Dashboard />, store)

    expect(wrapper.find(CircularProgress).length).toBe(1)
  })
})
