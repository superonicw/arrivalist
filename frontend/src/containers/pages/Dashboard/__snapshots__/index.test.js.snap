// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard page test renders correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [MockFunction],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [MockFunction],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Dashboard />
</ContextProvider>
`;
